File: typetwo_flutter/lib/services/date_converter.dart
// Vietnamese uses D/M/YYYY; Chinese uses YYYY/M/D or M/D.
// Applied as post-processing on translated output.
class DateConverter {
  static const _dateLangs = {'越南文'};

  // Chinese 年月日 characters left untranslated in output
  static final _reYmdZh = RegExp(r'(\d{4})年(\d{1,2})月(\d{1,2})日');
  static final _reMdZh = RegExp(r'(\d{1,2})月(\d{1,2})日');
  // YYYY/M/D → D/M/YYYY
  static final _reYmdSlash =
      RegExp(r'(?<!\d)(\d{4})/(\d{1,2})/(\d{1,2})(?!\d)');
  // YYYY-M-D → D-M-YYYY
  static final _reYmdDash =
      RegExp(r'(?<!\d)(\d{4})-(\d{1,2})-(\d{1,2})(?!\d)');
  // M/D where day > 12 (unambiguous)
  static final _reMdSlash = RegExp(r'(?<!\d)(\d{1,2})/(\d{1,2})(?!\d)');

  static String apply(String text, String targetLang) {
    if (!_dateLangs.contains(targetLang)) return text;
    text = text.replaceAllMapped(
        _reYmdZh, (m) => 'Ngày ${m[3]} tháng ${m[2]} năm ${m[1]}');
    text = text.replaceAllMapped(
        _reMdZh, (m) => 'ngày ${m[2]} tháng ${m[1]}');
    text = text.replaceAllMapped(
        _reYmdSlash, (m) => '${m[3]}/${m[2]}/${m[1]}');
    text = text.replaceAllMapped(
        _reYmdDash, (m) => '${m[3]}-${m[2]}-${m[1]}');
    text = text.replaceAllMapped(_reMdSlash, _swapMd);
    return text;
  }

  static String _swapMd(Match m) {
    final mo = int.parse(m[1]!);
    final d = int.parse(m[2]!);
    if (mo >= 1 && mo <= 12 && d >= 13 && d <= 31) return '${m[2]}/${m[1]}';
    return m[0]!;
  }
}

File: src/date_converter.py
import re

# Vietnamese uses D/M/YYYY; Chinese uses YYYY/M/D or M/D.
# Applied as post-processing on translated output.

_DATE_LANGS = {"越南文"}

# YYYY/M/D or YYYY-M-D → D/M/YYYY or D-M-YYYY
_RE_YMD_SLASH = re.compile(r'(?<!\d)(\d{4})/(\d{1,2})/(\d{1,2})(?!\d)')
_RE_YMD_DASH = re.compile(r'(?<!\d)(\d{4})-(\d{1,2})-(\d{1,2})(?!\d)')

# M/D where day > 12 (unambiguous: day can't be a month)
_RE_MD_SLASH = re.compile(r'(?<!\d)(\d{1,2})/(\d{1,2})(?!\d)')

# Chinese 年月日 characters left untranslated in output
_RE_YMD_ZH = re.compile(r'(\d{4})年(\d{1,2})月(\d{1,2})日')
_RE_MD_ZH = re.compile(r'(\d{1,2})月(\d{1,2})日')


def _swap_md(m: re.Match) -> str:
    mo, d = int(m.group(1)), int(m.group(2))
    if 1 <= mo <= 12 and 13 <= d <= 31:
        return f"{m.group(2)}/{m.group(1)}"
    return m.group(0)


def apply_date_conversion(text: str, target_lang: str) -> str:
    if target_lang not in _DATE_LANGS:
        return text
    text = _RE_YMD_ZH.sub(lambda m: f"Ngày {m.group(3)} tháng {m.group(2)} năm {m.group(1)}", text)
    text = _RE_MD_ZH.sub(lambda m: f"ngày {m.group(2)} tháng {m.group(1)}", text)
    text = _RE_YMD_SLASH.sub(lambda m: f"{m.group(3)}/{m.group(2)}/{m.group(1)}", text)
    text = _RE_YMD_DASH.sub(lambda m: f"{m.group(3)}-{m.group(2)}-{m.group(1)}", text)
    text = _RE_MD_SLASH.sub(_swap_md, text)
    return text
